analysis/metric_space.lean

```diff
@@ -277,6 +287,44 @@ theorem totally_bounded_of_metric {s : set α} :
                ⟨t, ft, h⟩ := H ε ε0 in
   ⟨t, ft, subset.trans h $ Union_subset_Union $ λ y, Union_subset_Union $ λ yt z, hε⟩⟩
 
+/--A metric space space is totally bounded if one can reconstruct up to any ε>0 any element of the
+space from finitely many data.-/
+lemma totally_bounded_of_finite_discretization {α : Type u} [metric_space α] (s : set α)
+  (H : ∀ε > (0 : real), ∃(β : Type*) (t : set β) (F : α → β),
+     finite t ∧ (∀x ∈ s, F x ∈ t) ∧ (∀x y ∈ s, F x = F y → dist x y < ε)) :
+  totally_bounded s :=
+have A : s = ∅ → totally_bounded s :=
+begin
+  assume hs,
+  rw totally_bounded_of_metric,
+  intros ε εpos,
+  exact ⟨∅, ⟨finite_empty, begin rw hs, by simp end⟩⟩
+end,
+have B : s ≠ ∅ → totally_bounded s :=
+begin
+  assume hs,
+  rcases exists_mem_of_ne_empty hs with ⟨x0, hx0⟩,
+  haveI : inhabited α := ⟨x0⟩,
+  rw totally_bounded_of_metric,
+  intros ε εpos,
+  rcases H ε εpos with ⟨β, t, F, ⟨finite_t, hFt, hF⟩⟩,
+  let Finv := function.inv_fun_on F s,
+  let t' := Finv '' t,
+  have : finite t' := finite_image _ finite_t,
+  have : s ⊆ ⋃ (y : α) (H : y ∈ t'), ball y ε :=
+  begin
+    assume x x_in_s,
+    let x' := Finv (F x),
+    have : x' ∈ s := function.inv_fun_on_mem ⟨x, x_in_s, rfl⟩,
+    have : F x' = F x := function.inv_fun_on_eq ⟨x, x_in_s, rfl⟩,
+    have : dist x x' < ε := hF _ _ ‹x ∈ s› ‹x' ∈ s› (this.symm),
+    simp,
```

Comment:
Having a non-finishing `simp` is a bit dangerous. Could you change this into a `simp only [...]`? You can replace it with `squeeze_simp` to get a message that tells you what to put in place of the `[...]`.
Also, `simp` followed by `exact foo` can usually be changed into `simpa using foo`, which is less dangerous.